packages/plugins/apollo-usage-report/src/reporter.ts

```diff
@@ -0,0 +1,189 @@
+      maxBatchDelay: options.maxBatchDelay ?? 20 * 60 * 1000, // 20min
+      maxBatchUncompressedSize: options.maxBatchUncompressedSize ?? 4 * 1024 * 1024, // 4mb
+      maxTraceSize: options.maxTraceSize ?? 10 * 1024 * 1024, // 10mb
+      exportTimeout: options.exportTimeout ?? 30 * 1000, // 60s
+    };
+    this.#reportHeaders = {
+      graphRef: getGraphRef(options),
+      hostname: options.hostname ?? getEnvVar('HOSTNAME') ?? '',
+      uname: options.uname ?? '', // TODO: find a cross-platform way to get the uname
+      runtimeVersion: options.runtimeVersion ?? '',
+      agentVersion: options.agentVersion || `graphql-yoga@${yoga.version}`,
+    };
+  }
+
+  addTrace(schemaId: string, options: Parameters<OurReport['addTrace']>[0]) {
+    const report = this.#getReport(schemaId);
+    report.addTrace(options);
+
+    if (
+      this.#options.alwaysSend ||
+      report.sizeEstimator.bytes >= this.#options.maxBatchUncompressedSize!
+    ) {
+      return this.sendReport(schemaId);
+    }
+
+    this.#nextSendAfterDelay ||= setTimeout(() => this.flush(), this.#options.maxBatchDelay);
+
+    return;
+  }
+
+  async flush() {
+    return Promise.allSettled([
+      ...this.#sending, // When flushing, we want to also wait for previous traces to be sent, because it's mostly used for clean up
+      ...Object.keys(this.#reportsBySchema).map(schemaId => this.sendReport(schemaId)),
+    ]);
+  }
+
+  async sendReport(schemaId: string) {
+    const sending = this.#sendReport(schemaId);
+    this.#sending.push(sending);
+    sending.finally(() => this.#sending?.filter(p => p !== sending));
+    return sending;
+  }
+
+  async #sendReport(schemaId: string) {
+    const {
+      fetchAPI: { fetch, CompressionStream, ReadableStream },
+    } = this.#yoga;
+    const report = this.#reportsBySchema[schemaId];
+    if (!report) {
+      throw new Error(`No report to send for schema ${schemaId}`);
+    }
+
+    if (this.#nextSendAfterDelay != null) {
+      clearTimeout(this.#nextSendAfterDelay);
+    }
+
+    delete this.#reportsBySchema[schemaId];
+    report.endTime = dateToProtoTimestamp(new Date());
+    report.ensureCountsAreIntegers();
+
+    const validationError = Report.verify(report);
+    if (!validationError) {
+      throw new TypeError(`Invalid report: ${validationError}`);
+    }
+
+    const encodedReport = Report.encode(report).finish();
+    const compressedReport = new ReadableStream({
+      start(controller) {
+        controller.enqueue(encodedReport);
+        controller.close();
+      },
+    }).pipeThrough(new CompressionStream('gzip'));
+
+    const { apiKey = getEnvVar('APOLLO_KEY'), endpoint = DEFAULT_REPORTING_ENDPOINT } =
+      this.#options;
+
+    for (let tries = 0; tries < 5; tries++) {
+      try {
+        const abortCtl = new AbortController();
+        this.#logger?.debug(`Sending report (try ${tries}/5)`);
+        const timeout = setTimeout(() => abortCtl.abort(), this.#options.exportTimeout);
```

Comment:
You can use `AbortSignal.timeout` here, maybe?

Reply:
I've never used this api, is there any way to "cancel" this timeout ? Or we just forget about it once the request have been fullfiled ?